pkg/providers/launchtemplate/launchtemplate.go

```diff
@@ -0,0 +1,190 @@
+/*
+Copyright 2024 The CloudPilot AI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package launchtemplate
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	ecs "github.com/alibabacloud-go/ecs-20140526/v4/client"
+	util "github.com/alibabacloud-go/tea-utils/v2/service"
+	"github.com/alibabacloud-go/tea/tea"
+	"github.com/patrickmn/go-cache"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/utils/pretty"
+
+	"github.com/cloudpilot-ai/karpenter-provider-alicloud/pkg/apis/v1alpha1"
+	"github.com/cloudpilot-ai/karpenter-provider-alicloud/pkg/operator/options"
+	"github.com/cloudpilot-ai/karpenter-provider-alicloud/pkg/providers/securitygroup"
+	"github.com/cloudpilot-ai/karpenter-provider-alicloud/pkg/providers/vswitch"
+	"github.com/cloudpilot-ai/karpenter-provider-alicloud/pkg/utils"
+)
+
+type Provider interface {
+	EnsureAll(context.Context, *v1alpha1.ECSNodeClass, *karpv1.NodeClaim,
+		[]*cloudprovider.InstanceType, string, map[string]string) ([]*LaunchTemplate, error)
+	DeleteAll(context.Context, *v1alpha1.ECSNodeClass) error
+	InvalidateCache(context.Context, string, string)
+}
+
+type LaunchTemplate struct {
+	Name          string
+	InstanceTypes []*cloudprovider.InstanceType
+	ImageID       string
+}
+
+type DefaultProvider struct {
+	sync.Mutex
+	region string
+	ecsapi ecs.Client
+	cache  *cache.Cache
```

Comment:
The cache is used to save the launchtemplate that has been created by karpenter, so that multiple nodeclaims can reuse the launchtemplate and reduce API calls.